### PR TITLE
First issues

### PR DESCRIPTION
**Pull request by perryr16**
1. Visitors cannot bookmark videos, instead, they see a fake button and a message telling them to login
2. Visitors cannot see tutorials that have `classroom: true` in database. This only works on main screen at moment. They could artificially type in `/turorials/1` but there is no button to allow them to get there
3. Deleting a tutorial deletes its videos

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -22,7 +22,7 @@ def destroy
 
   def update
     github_token = request.env['omniauth.auth'][:credentials][:token]
-    current_user.update_attributes(github_token: github_token)
+    current_user.update(github_token: github_token)
```
Comment (perryr16) [inline app/controllers/sessions_controller.rb:25]: rubocop won this battle

File: app/controllers/welcome_controller.rb [hunk 2]
```diff
@@ -1,10 +1,5 @@
 class WelcomeController < ApplicationController
   def index
-    if params[:tag]
-      tagged_tutorials = Tutorial.tagged_with(params[:tag])
-      @tutorials = tagged_tutorials.paginate(page: params[:page], per_page: 5)
-    else
-      @tutorials = Tutorial.all.paginate(page: params[:page], per_page: 5)
-    end
+    @tutorials = Tutorial.filtered_tutorials(params, current_user)
```
Comment (perryr16) [inline app/controllers/welcome_controller.rb:3]: moved all this to the model where I called a few new methods

File: app/models/tutorial.rb [hunk 3]
```diff
@@ -1,8 +1,28 @@
...
 
+  def self.filtered_tutorials(params, user)
+    tutorials = tagged_tutorials(params)
+    classroom_tutorials(user, tutorials)
+  end
+
+  def self.classroom_tutorials(user, tutorials)
+    return tutorials.where(classroom: false) unless user
+
+    tutorials
```
Comment (perryr16) [inline app/models/tutorial.rb:14]: This will only show tutorials with classroom: false, so visitors can see

File: app/models/tutorial.rb [hunk 4]
```diff
@@ -1,8 +1,28 @@
...
+  end
+
+  def self.tagged_tutorials(params)
+    if params[:tag]
+      tagged_tutorials = Tutorial.tagged_with(params[:tag])
+      tagged_tutorials.paginate(page: params[:page], per_page: 5)
+    else
+      Tutorial.all.paginate(page: params[:page], per_page: 5)
+    end
```
Comment (perryr16) [inline app/models/tutorial.rb:23]: This was pulled from the controller, put here to keep the controller skinny

File: app/models/user.rb [hunk 5]
```diff
@@ -3,8 +3,12 @@ class User < ApplicationRecord
   has_many :videos, through: :user_videos
 
   validates :email, uniqueness: true, presence: true
-  validates :password, presence: true
+  validates :password, presence: true, on: :update, allow_blank: true
```
Comment (perryr16) [inline app/models/user.rb:9]: kept having issues with the user unable to be updated.

File: app/models/user.rb [hunk 6]
```diff
@@ -3,8 +3,12 @@ class User < ApplicationRecord
...
   validates :first_name, presence: true
   enum role: { default: 0, admin: 1 }
   has_secure_password
+
+  def bookmarked_videos
+    videos.order(:tutorial_id, :position)
+  end
```
Comment (perryr16) [inline app/models/user.rb:17]: whoa, look at that AR call

File: app/views/tutorials/show.html.erb [hunk 7]
```diff
@@ -19,7 +19,8 @@
       <% if current_user %>
       <%= button_to "Bookmark", user_videos_path(video_id: @facade.current_video.id, user_id: current_user.id), method: :post, class: "btn btn-outline mb1 teal" %>
       <% else %>
-      <%= link_to "Bookmark", login_path, class: "btn btn-outline mb1 teal" %>
+      <p class="btn btn-outline mb1 teal" %> Bookmark </p>
```
Comment (perryr16) [inline app/views/tutorials/show.html.erb:22]: its a paragraph, but looks like a button

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -10,6 +10,15 @@
   </ul>
   <section>
     <h1>Bookmarked Segments</h1>
+    <% if current_user.videos %>
+    <ul>
+     <%current_user.bookmarked_videos.each do |video| %>
+      <li><%= link_to video.title, "/tutorials/#{video.tutorial_id}/?video_id=#{video.id}" %></li>
+     <% end %>
```
Comment (perryr16) [inline app/views/users/show.html.erb:17]: It looks like this PR includes code from my bookmarked PR. we'll have to be careful

File: spec/services/youtube_service_spec.rb [hunk 9]
```diff
@@ -15,7 +15,7 @@
 
     it "playlist_items_info" do
       playlist_items_info = @service.playlist_items_info(@id)
-      expect(playlist_items_info[:items].length).to eq(13)
+      expect(playlist_items_info[:items].length).to eq(10)
```
Comment (perryr16) [inline spec/services/youtube_service_spec.rb]: This kept failing. the number of videos in the playlist has already jumped to 15. I changed it to a turning playlist that has 10 vidoes

**Review by perryr16 — COMMENTED** [review]
many small issues

File: app/models/user.rb (hunk 5 above)
Comment (stellakunzang) [reply app/models/user.rb:9]: okie dokie

**Review by stellakunzang — APPROVED** [review]
none of my gifs will work here so now I have to use dumb words to say "Great Work!"

**Pull request merged**